src/main/java/org/apache/solr/handler/dataimport/SolrCloudWriter.java

```diff
@@ -0,0 +1,150 @@
+ */
+package org.apache.solr.handler.dataimport;
+
+import org.apache.solr.client.solrj.impl.Http2SolrClient;
+import org.apache.solr.client.solrj.request.AbstractUpdateRequest;
+import org.apache.solr.client.solrj.request.UpdateRequest;
+import org.apache.solr.cloud.ZkController;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.SolrInputDocument;
+import org.apache.solr.common.cloud.ClusterState;
+import org.apache.solr.common.cloud.DocCollection;
+import org.apache.solr.common.cloud.Slice;
+import org.apache.solr.common.cloud.ZkCoreNodeProps;
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.apache.solr.core.CoreContainer;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.update.AddUpdateCommand;
+import org.apache.solr.update.SolrCmdDistributor;
+import org.apache.solr.update.processor.UpdateRequestProcessor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.invoke.MethodHandles;
+import java.util.Collections;
+import java.util.Map;
+import java.util.function.Consumer;
+
+
+@SuppressWarnings("unused")
+public class SolrCloudWriter extends SolrWriter { //not sure about ascendant
+
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  public static final String DST_COLL_PARAM = "destination-collection";
+  private final Http2SolrClient updateClient;
+  private final String destColl;
+  private final DocCollection destDocColl;
+  private final SolrCmdDistributor solrCmdDistributor;
+
+  @SuppressWarnings("unused")
+  public SolrCloudWriter(UpdateRequestProcessor processor, SolrQueryRequest req) {
+    super(processor, req);
+    updateClient = req.getCoreContainer().getUpdateShardHandler().getUpdateOnlyHttpClient();
+    destColl = req.getParams().get(DST_COLL_PARAM);
+    if (destColl == null) {
+      throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, SolrWriter.class.getCanonicalName() + " requires " + DST_COLL_PARAM);
+    }
+    CoreContainer cc = req.getCoreContainer();
+    ZkController zkController = cc.getZkController();
+    ClusterState clusterState = zkController.getClusterState();
+    destDocColl = clusterState.getCollection(destColl);
+
+    solrCmdDistributor = new SolrCmdDistributor(req.getCoreContainer().getUpdateShardHandler());
+  }
+
+  @Override
+  public void close() {
+    try {
+      super.close(); // some legacy housekeeping
+    } finally {
+      solrCmdDistributor.close();
+    }
+  }
+
+  @Override
+  public boolean upload(SolrInputDocument doc) {
+    try {
+      AddUpdateCommand command = new AddUpdateCommand(req);// accesses schema !
+      command.solrDoc = doc;
+      Slice slice = destDocColl.getRouter().getTargetSlice(command.getIndexedIdStr(), doc, null, req.getParams(), destDocColl);
+      SolrCmdDistributor.StdNode shardLeaderNode = new SolrCmdDistributor.StdNode(new ZkCoreNodeProps(slice.getLeader()), destColl, slice.getName());
```

Comment:
Why do we need to use SolrCmdDistributor. Why can't we use SolrClient ?

Reply:
I'm not sure, but here I: 

1. target shard leader (don't know if the client does it), and 
2. submit updates asynchronously (note SolrWriter interface accepts docs one by one). 
 
For low freq ops, I use update client. Just explaining, not arguing.

Reply:
> target shard leader (don't know if the client does it)

Client always sends updates to leader

> submit updates asynchronously (note SolrWriter interface accepts docs one by one).

But the actual update is done one at a time. Not a huge improvement. However we can have a separate thread to update

Reply:
> we can have a separate thread to update

That's what I'm trying to avoid reusing ones from SolrCmdDistributor.

Reply:
I think we can commit this now and may be optimize it later